R: Arrow timestamps in seconds, milliseconds and nanoseconds

james-atkins opened the pull request:
Fixes duckdb#2734

Currently only Arrow timestamps in microseconds are supported by the R
package - for other units, an error is raised. Add support for
timestamps in seconds, milliseconds and nanoseconds.

All units are converted to R's POSIXct format. As there may be a
precision loss when nanoseconds are converted, a one-time warning is
shown.

Mytherin commented on tools/rpkg/src/statement.cpp:
```diff
@@ -268,20 +274,67 @@ static void transform(Vector &src_vec, SEXP &dest, idx_t dest_offset, idx_t n) {
 		VectorToR<int32_t, uint32_t>(src_vec, n, INTEGER_POINTER(dest), dest_offset, NA_INTEGER);
 		break;
+	case LogicalTypeId::TIMESTAMP_SEC: {
+		auto src_data = FlatVector::GetData<int64_t>(src_vec);
```
This code seems to be duplicated four times, could we extract this into a templated function?

For example, see how this is done [in the Python wrapper](https://github.com/duckdb/duckdb/blob/b48ff00354c88ea9ce4c9554adfa10689146cdd2/tools/pythonpkg/src/array_wrapper.cpp#L664).

Mytherin commented:
Thanks for the PR! Looks good. One comment:

Mytherin commented on tools/rpkg/src/statement.cpp:
```diff
@@ -268,20 +274,67 @@ static void transform(Vector &src_vec, SEXP &dest, idx_t dest_offset, idx_t n) {
+		SET_CLASS(dest, RStrings::get().POSIXct_POSIXt_str);
+		Rf_setAttrib(dest, RStrings::get().tzone_sym, RStrings::get().UTC_str);
+
+		static bool precision_warning = false;
```
This isn't thread safe, right? Does it matter or is this code guaranteed to only be called once because of the R interpreter? @hannesmuehleisen thoughts on this?

Mytherin replied:
Could use [`std::call_once`](https://en.cppreference.com/w/cpp/thread/call_once) for this.

james-atkins commented:
Thanks for the comments. Will tidy up the duplicated code and look at call_once.

Looking at the Python wrapper, it seems that I should handle TIMESTAMP_TZ too? And then presumably set the "correct" timezone in the returned R type. Any pointers as to what the correct timezone is for a TIMESTAMP_TZ vs a TIMESTAMP, and how do I get it?

Alex-Monahan commented:
I'm not sure where the current timezone is stored on the C++ side, but here is a way to get it using SQL in case that is helpful as a starting point!
[Timestamp docs](https://duckdb.org/docs/sql/data_types/timestamp)

```sql
SELECT * FROM duckdb_settings() WHERE name = 'TimeZone';
-- Americas/Los_Angeles
```

james-atkins committed:
R package: timestamps in seconds, milliseconds and nanoseconds

Add support for Arrow datasets with timestamps in seconds, milliseconds
and nanoseconds.

All resolutions are converted to R's POSIXct format. As there may be a
precision loss when nanoseconds are converted, a one-time warning is
shown.

Fixes #2734

Mytherin reviewed (approved):
Thanks for the updates! This looks good. Ready for merge after feature freeze ends.
